File: 05Feb2025/backend_py/routes/allocations.py
""" 
Module contains the routes for the allocations API.

Endpoints:
    - GET /allocations: Get all allocations.
    - POST /allocations: Allocate a book to a member.
    - DELETE /allocations/{book_isbn}/{member_id}: Deallocate a book from a member.
"""

from fastapi import APIRouter, HTTPException
from fastapi.encoders import jsonable_encoder
from models import AllocationDetails
from database import loadData, saveData, ALLOCATIONS_FILE, BOOKS_FILE

router = APIRouter(prefix="/allocations", tags=["Allocations"])

@router.get("/")
def getAllocations():
    """
    Get all allocations.
    """
    allocations = loadData(ALLOCATIONS_FILE)
    return allocations["allocations"]

@router.post("/")
def allocateBook(allocation_data: AllocationDetails):
    """
    Allocate a book to a member.

    Args:
        allocation_data (AllocationDetails): Details of the book allocation.

    Returns:
        dict: Success message or error message.
    """
    books = loadData(BOOKS_FILE)
    allocations = loadData(ALLOCATIONS_FILE)

    # Find the book by ISBN
    book = next((b for b in books["books"] if b["isbn"] == allocation_data.book_isbn), None)
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")

    if book["allocated_copies"] >= book["total_copies"]:
        raise HTTPException(status_code=400, detail="All copies of the book are already allocated")

    # Check if the book is already allocated to the same member
    for allocated_book in allocations["allocations"]:
        if allocated_book["book_isbn"] == allocation_data.book_isbn and allocated_book["user_id"] == allocation_data.user_id:
            raise HTTPException(status_code=400, detail="Book already allocated to this member")

    # Increase allocated copies
    book["allocated_copies"] += 1
    allocations["allocations"].append(jsonable_encoder(allocation_data))

    saveData(ALLOCATIONS_FILE, allocations)
    saveData(BOOKS_FILE, books)

    return {"message": "Book allocated successfully"}

@router.delete("/{book_isbn}/{member_id}")
def deallocateBook(book_isbn: str, member_id: int):
    """
    Deallocate a book from a member.

    Args:
        book_isbn (str): The ISBN of the book to deallocate.
        member_id (int): The ID of the member.

    Returns:
        dict: Success message.
    """
    books = loadData(BOOKS_FILE)
    allocations = loadData(ALLOCATIONS_FILE)

    # Find the book
    book = next((b for b in books["books"] if b["isbn"] == book_isbn), None)
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")

    # Filter out the allocation
    new_allocations = [
        allocation for allocation in allocations["allocations"]
        if not (allocation["book_isbn"] == book_isbn and allocation["user_id"] == member_id)
    ]

    if len(new_allocations) == len(allocations["allocations"]):
        raise HTTPException(status_code=400, detail="No such allocation found")

    # Reduce allocated copies
    book["allocated_copies"] -= 1

    allocations["allocations"] = new_allocations
    saveData(ALLOCATIONS_FILE, allocations)
    saveData(BOOKS_FILE, books)

    return {"message": "Book deallocated successfully"}

File: 05Feb2025/backend_js/routes/allocations.js
/**
 * Express router providing allocation related routes.
 * @module routes/allocations
 * @requires express.Router()
 * @requires file_handling
 */

const express = require('express');
const router = express.Router();
const { loadFile, saveFile, ALLOCATIONS_FILE, MEMBERS_FILE, BOOKS_FILE } = require('../file_handling');

/**
 * POST /allocations
 * Allocates a book to a member.
 * @name addAllocation
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.post('/', async (req, res) => {
    try {
        const { user_id, book_isbn, book_name, from_date, to_date } = req.body;
        const membersData = await loadFile(MEMBERS_FILE);
        const booksData = await loadFile(BOOKS_FILE);
        const allocationsData = await loadFile(ALLOCATIONS_FILE);

        const member = membersData.members.find(m => m.id === parseInt(user_id));
        if (!member) {
            return res.status(404).json({ error: 'Member not found' });
        }

        const book = booksData.books.find(b => b.isbn === book_isbn);
        if (!book) {
            return res.status(404).json({ error: 'Book not found' });
        }

        if (book.allocated_copies >= book.total_copies) {
            return res.status(400).json({ error: 'No copies available for allocation' });
        }

        // Check if the book is already allocated to this member
        const existingAllocation = allocationsData.allocations.find(entry => entry.book_isbn === book_isbn && entry.user_id === parseInt(user_id));
        if (existingAllocation) {
            return res.status(400).json({ error: 'Book already allocated to this member' });
        }

        // Allocate book
        allocationsData.allocations.push({ user_id: parseInt(user_id), book_isbn: book_isbn, book_name: book_name, from_date: from_date, to_date: to_date });
        book.allocated_copies += 1;

        await saveFile(ALLOCATIONS_FILE, allocationsData);
        await saveFile(BOOKS_FILE, booksData);
        res.json({ message: 'Book allocated successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to allocate book' });
    }
});

/**
 * DELET /allocations/:isbn/:memberId
 * Deallocates a book from a member.
 * @name deleteAllocation
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.delete('/:isbn/:memberId', async (req, res) => {
    try {
        const { isbn, memberId } = req.params;
        const membersData = await loadFile(MEMBERS_FILE);
        const booksData = await loadFile(BOOKS_FILE);
        const allocationsData = await loadFile(ALLOCATIONS_FILE);

        const member = membersData.members.find(m => m.id === parseInt(memberId));
        if (!member) {
            return res.status(404).json({ error: 'Member not found' });
        }

        const book = booksData.books.find(b => b.isbn === isbn);
        if (!book) {
            return res.status(404).json({ error: 'Book not found' });
        }

        // Remove the book from allocated list
        const prevAllocatedLength = allocationsData.allocations.length;
        allocationsData.allocations = allocationsData.allocations.filter(entry => entry.book_isbn !== isbn || entry.user_id !== parseInt(memberId));

        if (allocationsData.allocations.length === prevAllocatedLength) {
            return res.status(400).json({ error: 'Book not allocated to this member' });
        }

        // Reduce allocated copies count
        book.allocated_copies = Math.max(0, book.allocated_copies - 1);

        await saveFile(ALLOCATIONS_FILE, allocationsData);
        await saveFile(BOOKS_FILE, booksData);
        res.json({ message: 'Book deallocated successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to deallocate book' });
    }
});

/**
 * GET /allocations
 * Retrieves all book allocations.
 * @name getAllocations
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.get('/', async (req, res) => {
    try {
        const data = await loadFile(ALLOCATIONS_FILE);
        res.json(data.allocations);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load allocations' });
    }
});

module.exports = router;
